File: core/src/main/java/com/tp/holdem/core/model/Deck.java
package com.tp.holdem.core.model;


import io.vavr.collection.List;
import lombok.Data;

import java.util.stream.IntStream;

@Data
public class Deck {
	private java.util.List<Card> cards = List.of(Honour.values())
			.flatMap(honour ->
					List.of(
							Card.from(Suit.HEART, honour),
							Card.from(Suit.CLUB, honour),
							Card.from(Suit.DIAMOND, honour),
							Card.from(Suit.SPADE, honour)
					)
			).shuffle()
			.toJavaList();

	public void dealCards(final int numberOfCards, final List<Player> players) {
		final List<Player> playing = players
				.filter(Player::playing);

		IntStream.range(0, numberOfCards).forEach(i ->
				playing.forEach(player -> {
					player.addCard(cards.get(0));
					cards.remove(0);
				})
		);
	}

	public Card drawCard() {
		return List.ofAll(cards).headOption()
				.peek(card -> cards.remove(0))
				.getOrElseThrow(() -> new IllegalStateException("Card cannot be drawn - deck is empty"));
	}

}

File: model/src/main/kotlin/com/tp/holdem/model/Deck.kt
package com.tp.holdem.model


import com.tp.holdem.common.model.Honour
import com.tp.holdem.common.model.Suit
import io.vavr.collection.List

class Deck {
    companion object {
        @JvmStatic
        fun brandNew(): Deck {
            return Deck()
        }
    }

    private var cards = List.of(*Honour.values())
            .flatMap { honour ->
                List.of(
                        Card.from(Suit.HEART, honour),
                        Card.from(Suit.CLUB, honour),
                        Card.from(Suit.DIAMOND, honour),
                        Card.from(Suit.SPADE, honour)
                )
            }
            .shuffle()

    fun dealCards(numberOfCards: Int, players: List<Player>): List<Player> {
        return players
                .filter { it.playing() }
                .map { player -> player.withCards(List.fill(numberOfCards) { this.drawCard() }) }
                .appendAll(players.filter { it.notPlaying() })
    }

    fun drawCards(number: Int): List<Card> {
        val drawn = cards.take(number)
        cards = cards.removeAll(drawn)
        return drawn
    }

    fun drawCard(): Card {
        return cards.headOption()
                .peek { cards = cards.removeAt(0) }
                .getOrElseThrow { IllegalStateException("Card cannot be drawn - deck is empty") }
    }
}
